src/core/context/utils.ts

```diff
@@ -206,7 +206,295 @@ export function filterMessagesByLLMCapabilities(
+                        mimeType: anyObj.mimeType,
+                        filename: anyObj.filename,
+                    },
+                ];
+            }
+            // Generic object: remove huge base64 fields and stringify
+            const cleaned = sanitizeDeepObject(anyObj);
+            return JSON.stringify(cleaned);
+        }
+
+        // Fallback
+        return JSON.stringify(result ?? '');
+    } catch (err) {
+        logger.warn(`sanitizeToolResultToContent failed, falling back to string: ${String(err)}`);
+        try {
+            return JSON.stringify(result ?? '');
+        } catch {
+            return String(result ?? '');
+        }
+    }
+}
+
+/**
+ * Produce a short textual summary for tool content, to be used with providers
+ * that only accept text for tool messages (e.g., OpenAI/Anthropic tool role).
+ */
+export function summarizeToolContentForText(content: InternalMessage['content']): string {
+    if (!Array.isArray(content)) return String(content || '');
+    const parts: string[] = [];
+    for (const p of content) {
+        if (p.type === 'text') {
+            parts.push(p.text);
+        } else if (p.type === 'image') {
+            // Try estimating size
+            let bytes = 0;
+            if (typeof p.image === 'string') bytes = Math.floor((p.image.length * 3) / 4);
+            else if (p.image instanceof ArrayBuffer) bytes = p.image.byteLength;
+            else if (p.image instanceof Uint8Array) bytes = p.image.length;
+            else if (p.image instanceof Buffer) bytes = p.image.length;
+            parts.push(`[image ${p.mimeType || 'image'} ~${Math.ceil(bytes / 1024)}KB]`);
+        } else if (p.type === 'file') {
+            let bytes = 0;
+            if (typeof p.data === 'string') bytes = Math.floor((p.data.length * 3) / 4);
+            else if (p.data instanceof ArrayBuffer) bytes = p.data.byteLength;
+            else if (p.data instanceof Uint8Array) bytes = p.data.length;
+            else if (p.data instanceof Buffer) bytes = p.data.length;
+            const label = p.filename ? `${p.filename}` : `${p.mimeType || 'file'}`;
+            parts.push(`[file ${label} ~${Math.ceil(bytes / 1024)}KB]`);
+        }
+    }
+    const summary = parts.join('\n');
+    // Avoid passing enormous text anyway
+    return summary.slice(0, 4000);
+}
+
+// Helper: estimate base64 byte length from string length
+function base64LengthToBytes(charLength: number): number {
+    // 4 base64 chars -> 3 bytes; ignore padding for approximation
+    return Math.floor((charLength * 3) / 4);
+}
+
+/**
+ * Convert arbitrary tool content to safe text for providers that only accept textual tool messages.
+ * - If content is an array of parts, summarize it.
+ * - If content is a string that looks like base64/data URI, replace with a short placeholder.
+ * - Otherwise pass text through.
+ */
+export function toTextForToolMessage(content: InternalMessage['content'] | string | null): string {
```

Comment:
`| string | null` are already covered in `InternalMessage['content']` so this is redundant, we can just directly use `content: InternalMessage['content']`